DOC: Replace wikipedia links (#4853)
okta/okta-developer-docs: packages/@okta/vuepress-site/docs/concepts/saml/index.md

In this case, BigMart (who is providing this application) will need to take care of user authentication. The simple way is to require a different username and password from users working at JuiceCo. But think about all the users that this application will need to maintain - including all of the other suppliers and their users who need to access the application.

A more elegant way to solve this problem is to allow JuiceCo and every other supplier to share or "federate" the identities with BigMart. As an employee of JuiceCo, you already have a corporate identity and credentials. What Federated Identity provides is a secure way for the supermarket chain (Service Provider) to externalize authentication by integrating with the existing identity infrastructure of its suppliers (Identity Provider).

This type of use case is what led to the birth of federated protocols such as [Security Assertion Markup Language (SAML)](https://docs.oasis-open.org/security/saml/Post2.0/sstc-saml-tech-overview-2.0.html).

> See the [Security Assertion Markup Language (SAML) V2.0 Technical Overview](http://docs.oasis-open.org/security/saml/Post2.0/sstc-saml-tech-overview-2.0.html) for a more in-depth overview.

Reviewer: You don't need this line if you link to the standard in the line immediately above it.

Author: Good point. Thanks :)

- IdP Sign-in URL - This is the endpoint on the IdP side where SAML requests are posted. The SP needs to obtain this information from the IdP.

The easiest way to implement SAML is to leverage an open-source SAML toolkit. Find a list at the end of this article of recommended toolkits for several languages. These toolkits provide the logic needed to digest the information in an incoming SAML Response. In addition, if the SP needs to support the SP-initiated sign-in flow, the toolkits also provide the logic needed to generate an appropriate SAML authentication request.

### Single IdP vs multiple IdPs

If you’re building an internal integration and you want to SAML-enable it to integrate with your corporate SAML identity provider, then you’re looking at supporting only a single IdP. In this case, your integration only needs to deal with a single set of IdP metadata (cert, endpoints, and so on).

<div class="half">

![Single IdP](/img/saml/saml_guidance_one_idp.png)

</div>

> If you’re an independent software vendor (ISV) building an enterprise SaaS product, or if you’re building an external facing website/portal/community for your customers and partners, then you need to look at supporting multiple IdPs. This is the typical use case for many SaaS ISVs that need to integrate with customers' corporate identity infrastructure. Depending on the architecture of your application, you need to think about ways to store the SAML configuration (Certificates or IdP sign-in URLs, for example) from each identity provider, as well as how to provide the necessary SP information for each.

Reviewer 2: ```suggestion
Independent software vendors (ISV) building an enterprise SaaS product or an external facing website/portal/community for customers and partners should look at supporting multiple IdPs. This is the typical use case for many SaaS ISVs that need to integrate with customers' corporate identity infrastructure. Depending on the architecture of your application, think about ways to store the SAML configuration (Certificates or IdP sign-in URLs, for example) from each Identity Provider. You should also think about how to provide the necessary SP information for each.
```

<div class="half">

![Many IdPs](/img/saml/saml_guidance_many_idp.png)

</div>

> A key consideration involves the ACS URL endpoint on the SP side where SAML responses are posted. It’s possible to expose a single endpoint even when dealing with multiple IdPs. For a single-instance multi-tenant application where the tenancy isn't defined in the URL (such as when using a subdomain), this might be a simpler way to implement. However, you must then rely on additional information in the SAML response to determine which IdP is trying to authenticate (for example, using the IssuerID). If your application is set up in a multi-tenant fashion with domain information in the URL (for example, using either `https://domain1.example.com` or `https://www.example.com/domain1`), then having an ACS URL endpoint for each subdomain might be a good option since the URL itself identifies the domain.

Reviewer 2: ```suggestion
A key consideration involves the ACS URL endpoint on the SP side where SAML responses are posted. It’s possible to expose a single endpoint even when dealing with multiple IdPs. For a single-instance multi-tenant application where the tenancy isn't defined in the URL (such as when using a subdomain), this might be a simpler way to implement. However, you must then rely on additional information in the SAML response to determine which IdP is trying to authenticate (for example, using the IssuerID). Applications set up in a multi-tenant fashion with domain information in the URL (for example, using either `https://domain1.example.com` or `https://www.example.com/domain1`) can benefit from having an ACS URL endpoint for each subdomain. It's a good option since the URL itself identifies the domain.
```

### Understanding SP-initiated sign-in flow

> As discussed earlier, an IdP-initiated sign-in flow starts from the IdP. Since it begins on the IdP side, there’s no additional context about what the user is trying to access on the SP side other than the fact that the user is trying to get authenticated and access the SP. Typically, after the user is authenticated, the browser will be taken to a generic landing page in the SP.

Reviewer 2: ```suggestion
As discussed earlier, an IdP-initiated sign-in flow starts from the IdP. So, there’s no additional context about what the user is trying to access on the SP side. The only context is that the user is trying to get authenticated and access the SP. Typically, after the user is authenticated, the browser is redirected to a generic landing page in the SP.
```

> In an SP-initiated flow, the user tries to access a protected resource directly on the SP side without the IdP being aware of the attempt. Two issues arise. The first is the need to identify the right IdP if authentication of a federated identity is needed. With SP-initiated sign in, the SP initially doesn't know anything about the identity. As a developer, you need to figure out how the SP can determine which IdP should be receiving the SAML request. In some cases, if your application URLs contain subdomain information that is mapped to a unique tenant and IdP, then the resource link being hit is enough to identify the IdP. If this isn't the case, then you might need to prompt the end user for additional information from the end user such as user ID, email, or a company ID. You need something that allows the SP to identify which IdP the user attempting to access the resource belongs to. Remember, you’re only prompting for an identifier, not credentials. Okta also supports passing the identifier to the IdP with the parameter "LoginHint", so that the user doesn't need to input the identifier again when redirected to the IdP to sign in. For instruction to trigger Okta to send the "LoginHint" to IdP, see [Redirecting with SAML Deep Links](/docs/reference/api/idps#redirecting-with-saml-deep-links).

Reviewer 2: I would put a colon after "Two issues arise:" and then bullet the two issues. This is just buried in a huge paragraph. I know you didn't write it - but it's a lot for the user to consume. 

```suggestion
In an SP-initiated flow, the user tries to access a protected resource directly on the SP side without the IdP being aware of the attempt. Two issues arise. The first is the need to identify the right IdP if authentication of a federated identity is required. With SP-initiated sign-in flows, the SP initially doesn't know anything about the identity. As a developer, you need to figure out how the SP can determine which IdP should be receiving the SAML request. In some cases, if your application URLs contain subdomain information that's mapped to a unique tenant and IdP, then the resource link being hit is enough to identify the IdP. If this isn't the case, then you might need to prompt the end user for additional information such as user ID, email, or a company ID. You need something that allows the SP to identify which IdP the user attempting to access the resource belongs to. Remember, you’re only prompting for an identifier, not credentials. Okta also supports passing the identifier to the IdP with the parameter "LoginHint", so that the user doesn't need to input the identifier again when redirected to the IdP to sign in. For instruction to trigger Okta to send the "LoginHint" to IdP, see [Redirecting with SAML Deep Links](/docs/reference/api/idps#redirecting-with-saml-deep-links).
```

### Exposing SAML configuration in SP

> As discussed before, the SP needs the IdP configuration to complete the SAML setup. While many ISVs choose to do this through support and email, the better way is to expose a self-service administrator page for your customer's IT administrator to enable SAML. SAML supports metadata on both the IdP and SP side. One way to configure the IdP/SP relationship on the SP side is to build the ability to receive an IdP metadata file and the ability to generate an SP metadata file for consumption by the IdP. This is the preferred method.

Reviewer 2: ```suggestion
As discussed before, the SP needs the IdP configuration to complete the SAML setup. While many ISVs choose to do this through support and email, there is a better way. Expose a self-service administrator page for your customer's IT administrator to enable SAML. SAML supports metadata on both the IdP and SP side. One way to configure the IdP/SP relationship on the SP side is to build the ability to receive an IdP metadata file. Then, build the ability to generate an SP metadata file for consumption by the IdP. This is the preferred method.
```
